### ISSUE TITLE
JpaModule Wildfly/JBoss EAP integration broken

### ISSUE THREAD

**Issue by JochenReinhardt** [crnk-project/crnk-framework#467]
When running the Wildfly integration example in JBoss EAP 7.1 I get the following exception:
```
2019-01-31 07:44:25,620 ERROR [org.jboss.msc.service.fail] (ServerService Thread Pool -- 285) MSC000001: Failed to start service jboss.undertow.deployment.default-server.default-host./exa
mple: org.jboss.msc.service.StartException in service jboss.undertow.deployment.default-server.default-host./example: java.lang.IllegalStateException: resourceType 'schedule' already exis
ts, cannot add entry RegistryEntryImpl[type=schedule, path=schedule]
        at org.wildfly.extension.undertow.deployment.UndertowDeploymentService$1.run(UndertowDeploymentService.java:84)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
        at org.jboss.threads.JBossThread.run(JBossThread.java:320)
Caused by: java.lang.IllegalStateException: resourceType 'schedule' already exists, cannot add entry RegistryEntryImpl[type=schedule, path=schedule]
        at io.crnk.core.engine.internal.utils.PreconditionUtil.fail(PreconditionUtil.java:50)
        at io.crnk.core.engine.internal.utils.PreconditionUtil.verify(PreconditionUtil.java:131)
        at io.crnk.core.engine.registry.DefaultResourceRegistryPart.addEntry(DefaultResourceRegistryPart.java:43)
        at io.crnk.core.engine.internal.registry.ResourceRegistryImpl.addEntry(ResourceRegistryImpl.java:215)
        at io.crnk.core.module.ModuleRegistry.applyRepositoryRegistration(ModuleRegistry.java:559)
        at io.crnk.core.module.ModuleRegistry.applyRepositoryRegistrations(ModuleRegistry.java:461)
        at io.crnk.core.module.ModuleRegistry.init(ModuleRegistry.java:382)
        at io.crnk.core.boot.CrnkBoot.bootDiscovery(CrnkBoot.java:223)
        at io.crnk.core.boot.CrnkBoot.boot(CrnkBoot.java:196)
        at io.crnk.rs.CrnkFeature.configure(CrnkFeature.java:71)
```
I also tried with Wildfly 15.0.1.Final and I get a similar result:
```
09:03:57,729 ERROR [org.jboss.msc.service.fail] (ServerService Thread Pool -- 77) MSC000001: Failed to start service jboss.deployment.unit."example.war".undertow-deployment: org.jboss.msc.service.StartException in service jboss.deployment.unit."example.war".undertow-deployment: java.lang.IllegalStateException: resourceType 'users' already exists, cannot add entry RegistryEntryImpl[type=users, path=users]                                                                                                                                                          
        at org.wildfly.extension.undertow.deployment.UndertowDeploymentService$1.run(UndertowDeploymentService.java:81)                                                                    
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)                                                                                                         
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)                                                                                                                        
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)                                                                                
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1985)                                                                                                
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1487)                                                                                   
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1378)                                                                                         
        at java.lang.Thread.run(Thread.java:748)                                                                                                                                           
        at org.jboss.threads.JBossThread.run(JBossThread.java:485)                                                                                                                         
Caused by: java.lang.IllegalStateException: resourceType 'users' already exists, cannot add entry RegistryEntryImpl[type=users, path=users]                                                
        at io.crnk.core.engine.internal.utils.PreconditionUtil.fail(PreconditionUtil.java:50)                                                                                              
        at io.crnk.core.engine.internal.utils.PreconditionUtil.verify(PreconditionUtil.java:131)                                                                                           
        at io.crnk.core.engine.registry.DefaultResourceRegistryPart.addEntry(DefaultResourceRegistryPart.java:43)                                                                          
        at io.crnk.core.engine.internal.registry.ResourceRegistryImpl.addEntry(ResourceRegistryImpl.java:215)                                                                              
        at io.crnk.core.module.ModuleRegistry.applyRepositoryRegistration(ModuleRegistry.java:559)                                                                                         
        at io.crnk.core.module.ModuleRegistry.applyRepositoryRegistrations(ModuleRegistry.java:461)                                                                                        
        at io.crnk.core.module.ModuleRegistry.init(ModuleRegistry.java:382)                                                                                                                
        at io.crnk.core.boot.CrnkBoot.bootDiscovery(CrnkBoot.java:223)                                                                                                                     
        at io.crnk.core.boot.CrnkBoot.boot(CrnkBoot.java:196)                                                                                                                              
        at io.crnk.rs.CrnkFeature.configure(CrnkFeature.java:71)
```
I was able to make the deployment work by removing the repositories' `@Producer` methods and annotate the repositories directly with `@ApplicationScope`.
But then, the repositories seem to be not correctly initialized. An HTTP GET to `http://localhost:8080/example/api/schedule` results in:
```
{
    "errors": [
        {
            "status": "500",
            "title": "INTERNAL_SERVER_ERROR",
            "detail": "queryFactory not set"
        }
    ]
}
```
And the server log shows:
```
09:14:20,564 ERROR [io.crnk.core.engine.internal.http.JsonApiRequestProcessor] (default task-1) failed to process request, unknown exception thrown: java.lang.IllegalStateException: queryFactory not set                                                                                                                                                                            
        at io.crnk.core.engine.internal.utils.PreconditionUtil.fail(PreconditionUtil.java:50)                                                                                              
        at io.crnk.core.engine.internal.utils.PreconditionUtil.verify(PreconditionUtil.java:131)                                                                                           
        at io.crnk.jpa.JpaRepositoryConfig.getQueryFactory(JpaRepositoryConfig.java:65)                                                                                                    
        at io.crnk.jpa.JpaEntityRepositoryBase.findAll(JpaEntityRepositoryBase.java:82)                                                                                                    
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)                                                                                                                     
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)                                                                                                   
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)                                                                                           
        at java.lang.reflect.Method.invoke(Method.java:498)                                                                                                                                
        at org.jboss.weld.bean.proxy.AbstractBeanInstance.invoke(AbstractBeanInstance.java:38)                                                                                             
        at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:106)                                                                                                
        at io.crnk.core.repository.Repository$ResourceRegistryAware$ResourceRepositoryV2$1866110108$Proxy$_$$_WeldClientProxy.findAll(Unknown Source)                                      
        at io.crnk.core.engine.internal.repository.ResourceRepositoryAdapterImpl$2.invoke(ResourceRepositoryAdapterImpl.java:91)                                                           
        at io.crnk.core.engine.internal.repository.ResponseRepositoryAdapter$RepositoryRequestFilterChainImpl.doFilter(ResponseRepositoryAdapter.java:233)                                 
        at io.crnk.core.engine.internal.repository.ResourceRepositoryAdapterImpl.findAll(ResourceRepositoryAdapterImpl.java:101)                                                           
        at io.crnk.core.engine.internal.dispatcher.controller.CollectionGetController.handleAsync(CollectionGetController.java:42)                                                         
        at io.crnk.core.engine.internal.dispatcher.controller.BaseController.handle(BaseController.java:58)                                                                                
        at io.crnk.core.engine.internal.http.DocumentFilterChainImpl.doFilter(DocumentFilterChainImpl.java:28)                                                                             
        at io.crnk.jpa.JpaModule$4$1.call(JpaModule.java:407)                                                                                                                              
        at io.crnk.jpa.JpaModule$4$1.call(JpaModule.java:403)                                                                                                                              
        at io.crnk.cdi.internal.CdiTransactionRunnerImpl.doInTransaction(CdiTransactionRunnerImpl.java:13)                                                                                 
        at io.crnk.cdi.internal.CdiTransactionRunnerImpl$Proxy$_$$_WeldSubclass.doInTransaction$$super(Unknown Source)
```
I was able to set the query factory to `io.crnk.jpa.query.criteria.JpaCriteriaQueryFactory` in the repository constructor manually. That at least makes the repository work.
But I found out that other repository settings (like decorators) are not initialized. Seems the module setup is somewhat broken with wildfly and CDI.

**Comment by remmeier** [comment]
I cannot exactly say what the issue is, but everything seems related to service discovery/CDI.

- Make sure the repositories are only registered ones. So both @Producer and @ApplicationSsoped in parallel may trigger that depending on the CDI setup, not sure (not too familiiar).
- The missing JpaCriteriaQueryFactory  is a hint that the JPA Module is not setup or was not found. Is there a Producer in place? Maybe it would be good to add this information to "queryFactory not set".

**Comment by remmeier** [comment]
Yes @ApplicationSCoped should be the way to go similar to Spring. This way you can also inject the repositories in your own code and make use of it on the server side. 

not sure what the limitations of those weld proxies are. I would expect it to implement the same type. But it would be fairly straighforward to create an interface that implements the initialization contract of JpaEntityRepositoryBase. That for sure should work also with Weld then proxies get inbetween (which is not always the case, or rather strongly depends on the setup).

**Comment by JochenReinhardt** [comment]
I managed to get the initialization working by removing CDI annotations that will make weld create a proxy. Default scope (dependent) seems to be fine for the initialization as it runs only once, so there will be a single instance only anyway.

I'll give the new version a try anyways.

Now I am struggling with a many-to-many relation (based on an entity-relation). It was somehow working with version 2.9.20181028143330. Now I don't have an idea how to map it in my DTOs.

Do I need a custom relationship repository managing my m:n relation?
Should I create a new issue?

**Comment by remmeier** [comment]
where do you have an issue? feel free to create a dedicated ticket. Relationships should be handled automatically. In the past there was a JpaRelationshipRepository that has been replaced by the stock/default ForwardingRelationshipRepository from crnk-core. But for applications/consumers, both should be equal. The former can still be enabled should there be an issue in this area.

**Comment by JochenReinhardt** [comment]
I really would like to go with the new Forwarding kind of repositories.
I think I'll have to create a sample project with an actual many-to-many relation working with the legacy version, but failing with the new version. There is already too much clutter in the code I am currently working on. I 'll create a new issue as soon as I have the sample project up and running.

Thanks for the support so far.

**Comment by remmeier** [comment]
they hope with the forwarding one is to simplify everything quite a bit. That helped other places already and hopefully now with JPA as well as it allows to just focus on the resource repositories and get the relationships for free.

the other thing we did in our projects is too move away a bit from DTO mapping if not really necessary. rather go with a DB View or a couple of JsonIgnore or Transient annotations on the entiites. It helped to keep things simple. All to often DTOs and entities were almost equal and did not warrant the additional complexty. But DTO mapping or not, it should work equally well.

**Comment by JochenReinhardt** [comment]
There are other issues with the repository implementation.
As long as there are `final` methods in the classes created by CDI, like `JpaEntityRepositoryBase`, weld is complaining about these classes not being proxyable.

Do we really need the final methods?
Someone else might want to implement `io.crnk.jpa.JpaEntityRepositoryBase#findOne` differently, e.g. by using a cache.
Final methods are also an issue in testing. It just makes creating mocks harder. Would you mind removing the `final` from the methods?

**Comment by remmeier** [comment]
not strictly needed, but the reason it is so far there because people should override findAll(QuerySpec) and all other calls are delegated there. But can be removed if it causes issues. In general it might be better to create a producer and let it return an interface for the implementation since weld seems rather picky/limited :-(

**Comment by remmeier** [comment]
finals are gone
